source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -761,7 +781,13 @@ void Filter::onUpstreamConnection() {
   connecting_ = false;
   // Re-enable downstream reads now that the upstream connection is established
   // so we have a place to send downstream data to.
-  read_callbacks_->connection().readDisable(false);
+  if (!receive_before_connect_) {
+    read_callbacks_->connection().readDisable(false);
+  } else if (early_data_buffer_.length() > 0) {
+    getStreamInfo().getUpstreamBytesMeter()->addWireBytesSent(early_data_buffer_.length());
+    upstream_->encodeData(early_data_buffer_, early_data_end_stream_);
```

Comment:
Can you add a test that causes a high watermark on the upstream here, to make sure that any callbacks that re-renter this class have the correct state?

Reply:
Could you point to a test that is close enough to use for inspiration to what you mean here?